iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/mutex.hpp

```diff
@@ -17,81 +17,195 @@
-    ///         function did not returned any error.
-    ///        IMPORTANT! Unlocking and unlocked mutex is undefined behavior
-    ///         and the underlying c function will report success in this case!
-    bool unlock() noexcept;
-
-    /// @brief  Tries to lock the mutex object. If it is not possible to lock
-    ///         the mutex object try_lock will return an error. If the c
-    ///         function fails it will return false, otherwise true.
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    bool try_lock() noexcept;
-
-    /// @brief  Returns the native handle which then can be used in
-    ///         pthread_mutex_** calls. Required when a pthread_mutex_**
-    ///         call is not abstracted with this wrapper.
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    pthread_mutex_t get_native_handle() const noexcept;
-
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    pthread_mutex_t m_handle{};
+    Mutex(const Mutex&) = delete;
+    Mutex(Mutex&&) = delete;
+    Mutex& operator=(const Mutex&) = delete;
+    Mutex& operator=(Mutex&&) = delete;
+
+    /// @brief Locks the mutex.
+    /// @return When it fails it returns a MutexError describing the error.
+    ///         If the mutex is already locked it:
+    ///          * will be non blocking when the lock call comes from the same thread and the
+    ///              mutex type is MutexType::RECURSIVE
+    ///          * will be blocking when for all non MutexType::RECURSIVE
+    ///          * will be a MutexError::DEADLOCK_CONDITION when it is a
+    ///              MutexType::WITH_DEADLOCK_DETECTION
+
+    cxx::expected<MutexError> lock() noexcept;
+
+    /// @brief  Unlocks the mutex.
+    /// @return When it fails it returns a MutexError describing the error.
+    ///         If the mutex was locked by another thread and
+    ///         * MutexType::WITH_DEADLOCK_DETECTION it will return MutexError::NOT_OWNED_BY_THREAD
+    ///         * is not MutexType::WITH_DEADLOCK_DETECTION it may fail (depending on the platform)
+    ///         If the mutex was not locked it returns MutexType::NOT_OWNED_BY_THREAD
+    cxx::expected<MutexError> unlock() noexcept;
+
+    /// @brief Tries to lock the mutex.
+    /// @return If the lock was acquired MutexTryLock::LOCK_SUCCEEDED will be returned otherwise
+    ///         MutexTryLock::FAILED_TO_ACQUIRE_LOCK.
+    ///         If the lock is of MutexType::RECURSIVE the lock will also succeed.
+    ///         On failure it returns MutexError describing the failure.
+    cxx::expected<MutexTryLock, MutexError> try_lock() noexcept;
+
+  private:
+    Mutex() noexcept = default;
+
+  private:
+    friend class MutexBuilder;
+    friend class cxx::optional<Mutex>;
+
+    pthread_mutex_t m_handle;
+    bool m_isDescructable = true;
+};
+
+/// @todo iox-#1036 remove this, introduced to keep current API temporarily
```

Comment:
Introduced this alias to keep the current API and to not be forced to integrate this into iceoryx.